### PR TITLE
Tp table parsing

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Addressed #339, #338, #249

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by cseed** [comment]
I haven't looked at this yet, I have a new use case in seqr for the table code.  I need to be able to load data as `RDD[Annotation]` without pulling out a sample/variant key (or pull out a totally custom key).  Can I do that easily with the new interface?

**Comment by tpoterba** [comment]
Yes totally.  You'd just call ParsedLine.getValue.  getKey would return an
empty array, since you passed nothing in.

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
Updated and ready for review

**Comment by cseed** [comment]
Proposed interface changes:

class TextTableConfiguration

class TextTableReader

TextTableReader(conf)
TextTableReader(delimiter = "#", ...)

// only read fields 
TextTableReader.read(columnTypes: Map[String, Type], path: String, [select]): (TStruct, RDD[Annotation])

(and JSON)

for JSON:

JSONReader.read(t: Type, path: String): RDD[Annotation]

in expr language:

support

  Variant("chr:pos:ref:alt1,...,altN") 

(so Variant(v.toString) == v) and

  Variant(chr: String, pos: Int, ref: String, alts: Array[String])

Then we can do:

  annotatevariants table -v 'Variant(Chrom, Pos, Ref, Alts.split(","))'

  annotatevariants table -v 'Variant(Variant)'

To get this behavior, you'll have to build the EvalContext from the table type.

Add

  TStruct.filter(predicate: (Field) => Boolean): (TStruct, Filterer)

where

  type Filterer = (Annotation) => Annotation

This should make implementating importvariants table simple and elegant.

**Comment by tpoterba** [comment]
I mostly rewrote things to fit this interface, but found a pretty significant problem with it -- we lose our informative error messages.  Once we've got an `RDD[Annotation]`, we've lost our line context.  This new variant expression interface will be different and challenging with our users, and if there's a problem with the expression, I don't want to get a crash from a requirementError from the Variant constructor without any context, or a match error from a `Chr:Pos:Ref:Alt` with too few colon-split fields.

**Comment by tpoterba** [comment]
I also think a fixed result type of RDD will be a problem.  My CNV work involves parallelizing file parsing, and this interface wouldn't be compatible with that use case.

I think there are a lot of strengths of generating a `(String) => Annotation` parse function, and then letting the client code use that however it likes.  Including `sc.textFileLine(...)` in one read function prevents only a little bit of code duplication, and causes major headaches.

**Comment by cseed** [comment]
>  if there's a problem with the expression, I don't want to get a crash from a requirementError from the Variant constructor without any context

You need to do validation in the expr code.  User could isn't allowed to fail with a requirement error.

I'd solve the error message problem by carrying it along with the annotation.  Line can be generalized to carry line information about any type.

 > My CNV work involves parallelizing file parsing, and this interface wouldn't be compatible with that use case

I don't understand, can you elaborate on this?

**Comment by tpoterba** [comment]
What do you mean by validation?  

For the 'annotation line' are you suggesting a general error-catching wrapper?  I actually really like that, and I'll give it a go.

> CNV work
What I want to do with CNVs is something like  
```
val files: Array[String]
sc.paralellize(files)
  .map { f => readTable(f, config...) }
  .,map (convert to a hail better cnv representation)
```
Can't do that if readTable gives you an RDD

**Comment by tpoterba** [comment]
Here's my proposed interface (names to be changed, I'm terrible at those)

```
  case class WithSource[T](value: T, source: InputSource) {
    def map[U](f: T => U): WithSource[U] = {
      try {
        copy[U](value = f(value))
      } catch {
        case e: Exception => source.wrapError(e)
      }
    }
  }

  abstract class InputSource {
    def wrapError(e: Exception): Nothing
  }

  case class TextSource(line: String, file: String, position: Option[Int]) extends InputSource {
    def wrapError(e: Exception): Nothing = {
      val msg = e match {
        case _: FatalException => e.getMessage
        case _ => s"caught $e"
      }
      val lineToPrint =
        if (line.length > 62)
          line.take(59) + "..."
        else
          line

      log.error(
        s"""
           |$file${position.map(ln => ":" + (ln + 1)).getOrElse("")}: $msg
           |  offending line: $line""".stripMargin)
      fatal(
        s"""
           |$file${position.map(ln => ":" + (ln + 1)).getOrElse("")}: $msg
           |  offending line: $lineToPrint""".stripMargin)
    }
  }
```

**Comment by cseed** [comment]
> validation

Checking the arguments to the `Variant` constructor and generating nice error messages.

For CNV work, you want to parallelize over files and not lines within files?  You need to process the each file serially?

I'm not against having an additional interface like ParseContext that you can use both for the RDD interface and for the CNV stuff.  Another option might be to make `TableReader[C[_]](...): (TStruct, C[Annotation])` but you'll have to do some work to define a `C` that knows how to load itself from a file, for example.  It would be useful to be able to write code that can be used with either RDDs or local collections.

 > For the 'annotation line' are you suggesting a general error-catching wrapper?

Yep!  I'll look over your proposed interface.

Letting my mind wander a little here.  One of the challenges with Spark error handling is propagating errors from the workers back to the master.  RDDs are naturally used functionally, so functional error handling might be a better approach.  The `Try` monad is the normal way to do functional error handling in Scala.  Since we have concurrency, we have multiple errors and we want to preserve them all.  In addition, it would be nice if the new generalized `Try` monad tracked warnings (like VCFReport).

The main thing it isn't clear how to handle is writing RDDs.  You basically want a write to write out the values and return an RDD, but just with the errors and warnings, which you could transfer to the driver at the end with collect.

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
I'm now happy with the updates, ready for another look

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

File: src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVCF.scala [hunk 1]
```diff
@@ -28,22 +37,54 @@ object AnnotateVariantsVCF extends Command with VCFImporter {
...
+    val otherVds = {
+      val load = LoadVCF(vds.sparkContext, inputs.head, inputs, skipGenotypes = true)
+      if (options.split)
+        SplitMulti.run(state.copy(vds = load)).vds
+      else load
+    }
+
+    val multiallelicMsg =
+      """Merge behavior may not be as expected, as all alternate alleles are considered part of the key.  For example:
+        |  `22:140012:A:T,TTT' will not be annotated by `22:140012:A:T' or `22:140012:A:T'
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVCF.scala]: Do you want TTT in the second variant?

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVCF.scala]: indeed I do

File: src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVCF.scala [hunk 2]
```diff
@@ -28,22 +37,54 @@ object AnnotateVariantsVCF extends Command with VCFImporter {
...
+    (vds.wasSplit, otherVds.wasSplit) match {
+      case (true, true) =>
+      case (false, false) => warn(
+        s"""annotating a multiallelic vds from an unknown (unsplit) vcf
+            |  $multiallelicMsg""".stripMargin)
+      case (true, false) => warn(
+        s"""annotating a biallelic (split) vds from an unknown (unsplit) vcf
+            |  $multiallelicMsg""".stripMargin)
+      case (false, true) => warn(
+        s"""annotating a multiallelic vds from a biallelic (split) vcf
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVCF.scala]: Why are unsplit VDSes multiallelic and unsplit VCFs are unknown?

Also, VDS and VCF should be capitalized.

File: src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVDS.scala [hunk 3]
```diff
@@ -24,24 +27,49 @@ object AnnotateVariantsVDS extends Command {
...
+      case (None, Some(r)) => (false, r)
+      case _ => fatal("this module requires one of `--root' or `--code', but not both")
+    }
 
-    val readOtherVds = Read.run(state, Array("--skip-genotypes", "-i", filepath)).vds
+    val otherVds = Read.run(state, Array("--skip-genotypes", "-i", options.input)).vds
 
-    if (!readOtherVds.wasSplit)
-      fatal("cannot annotate from a multiallelic VDS, run `splitmulti' on that VDS first.")
+    val multiallelicMsg =
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AnnotateVariantsVDS.scala]: This logic is duplicated.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 4]
```diff
@@ -575,8 +609,21 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
           parseError(s"Got invalid argument `${a.`type`} to function `$fn'")
         TString
 
-      case ("hwe", rhs) =>
-        rhs.map(_.`type`) match {
+      case ("variant", _) =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: This should be upper case.

File: src/main/scala/org/broadinstitute/hail/expr/Type.scala [hunk 5]
```diff
@@ -104,6 +100,12 @@ abstract class Type extends BaseType {
       a => Option(a)
   }
 
+  def printPretty(compact: Boolean = false): String = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Type.scala]: print sounds like it prints.  Just overload pretty without a StringBuidler.  This should take a prettyAttrs flag, too.  Look at where `pretty` is called in the code and use this where you can.

File: src/main/scala/org/broadinstitute/hail/io/annotators/IntervalListAnnotator.scala [hunk 6]
```diff
@@ -39,57 +40,60 @@ object IntervalListAnnotator {
...
-              """.stripMargin)
-          }
-        }
-        })
-        .toArray
+              Interval(Locus(contig, start_str.toInt),
+                Locus(contig, end_str.toInt + 1))
+            case str => str.split("\t") match {
+              case Array(contig, start, end, direction, _) =>
+                assert(direction == "+" || direction == "-")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/io/annotators/IntervalListAnnotator.scala]: While you're editing this, can you make a nice error here?

File: src/main/scala/org/broadinstitute/hail/io/annotators/IntervalListAnnotator.scala [hunk 7]
```diff
@@ -39,57 +40,60 @@ object IntervalListAnnotator {
...
-    writeTextFile(filename, hConf) { fw =>
-      it.foreach { i =>
-        assert(i.start.contig == i.end.contig)
-        // interval list is 1-based, inclusive: [start, end]
-        fw.write(i.start.contig + ":" + i.start.position + "-" + (i.end.position - 1) + "\n")
-      }
+    writeTextFile(filename, hConf) {
+      fw =>
+        it.foreach {
+          i =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/io/annotators/IntervalListAnnotator.scala]: I like to have the `i =>` on the same line as the open brace.

File: src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    val firstLine = firstLines.head.value
+
+    val columns = if (noHeader) {
+      firstLine.split(separator, -1)
+        .zipWithIndex
+        .map { case (_, i) => s"_$i" }
+    }
+    else
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala]: I want to add a syntax checker that rejects PRs with dangling elses!

File: src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    config.commentChar, config.separator, config.missing, config.noHeader, config.noImpute)
+
+  def read(sc: SparkContext,
+    files: Array[String],
+    types: Map[String, Type] = TextTableConfiguration.defaultTypes,
+    commentChar: Option[String] = TextTableConfiguration.defaultComment,
+    separator: String = TextTableConfiguration.defaultSep,
+    missing: String = TextTableConfiguration.defaultMissing,
+    noHeader: Boolean = TextTableConfiguration.defaultNoHeader,
+    noImpute: Boolean = TextTableConfiguration.defaultNoImpute): (TStruct, RDD[WithContext[Annotation]]) = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala]: I don't like this duplication.  Is `TextTableReader.read(sc, files, TextTableConfiguration(param = override))` so bad?  Moreover, this function is never used, except with all default arguments, and you can make the other version have a default argument for the config.

File: src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  val doubleRegex = """^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$"""
+  val variantRegex = """^.+:\d+:[ATGC]+:([ATGC]+|\*)(,([ATGC]+|\*))*$"""
+  val headToTake = 20
+
+  def guessType(values: Seq[String], missing: String): Option[Type] = {
+    require(values.nonEmpty)
+
+    val size = values.size
+    val booleanMatch = values.exists(value => value.matches(booleanRegex))
+    val allBoolean = values.forall(value => value.matches(booleanRegex) || (value == missing))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala]: Might be cleaner to filter out the missing right at the beginning.

File: src/main/scala/org/broadinstitute/hail/expr/Type.scala (hunk 5 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/Type.scala]: Can't do this -- overloaded method with different return type doesn't work.  I've renamed it `toPrettyString` for now.

File: src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala (hunk 10 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala]: jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Pull request merged**